src/tracker/client.rs

```diff
@@ -0,0 +1,268 @@
+use super::*;
+use crate::common::*;
+
+pub(crate) struct Client {
+  peer_id: [u8; 20],
+  sock: UdpSocket,
+  state: State,
+}
+
+impl<'a> Client {
+  const UDP_TRACKER_MAGIC: u64 = 0x0000_0417_2710_1980;
+
+  pub fn new(peer_id: [u8; 20]) -> Result<Self> {
+    Ok(Self {
+      peer_id,
+      sock: UdpSocket::bind("0.0.0.0:0").context(error::Network)?,
+      state: State::Disconnected,
+    })
+  }
+
+  pub fn connect(&mut self, host_port: &HostPort) -> Result<()> {
+    let mut rng = rand::thread_rng();
+
+    self.sock.connect(host_port).context(error::Network)?;
+
+    let req = connect::Request {
+      protocol_id: Self::UDP_TRACKER_MAGIC,
+      action: 0x0000,
+      transaction_id: rng.gen(),
+    };
+
+    let mut buf = [0u8; mem::size_of::<connect::Response>()];
+    let (resp, _) = self.roundtrip(&req, &mut buf)?;
+
+    if resp.transaction_id != req.transaction_id || resp.action != req.action {
+      return Err(Error::TrackerResponse);
+    }
+
+    self.state = State::Connected {
+      id: resp.connection_id,
+      epoch: Instant::now(),
+      socket_addr: self.sock.peer_addr().context(error::Network)?,
+      host_port: host_port.clone(),
+    };
+
+    Ok(())
+  }
+
+  pub fn announce(&self, btinh: Infohash) -> Result<Vec<SocketAddr>> {
+    let mut rng = rand::thread_rng();
+    let req = announce::Request {
+      connection_id: self.state.get_connection_id()?,
+      action: 0x0001,
+      transaction_id: rng.gen(),
+      infohash: btinh.into(),
+      peer_id: self.peer_id,
+      downloaded: 0x0000,
+      left: u64::MAX,
+      uploaded: 0x0000,
+      event: 0x0000,
+      ip_address: 0x0000,
+      num_want: u32::MAX,
+      port: self.sock.local_addr().context(error::Network)?.port(),
+    };
+    let mut buf = [0u8; 8192]; // is there a BUFSIZ macro for Rust?
+    let (resp, len) = self.roundtrip(&req, &mut buf)?;
+
+    if resp.transaction_id != req.transaction_id || resp.action != req.action {
+      return Err(Error::TrackerResponse);
+    }
+    self.parse_compact_peer_list(&buf[mem::size_of::<announce::Response>()..len])
+  }
+
+  fn roundtrip<T: Request>(&self, req: &T, rxbuf: &mut [u8]) -> Result<(T::Response, usize)> {
```

Comment:
Instead of returning a `usize`, can `send_and_retry_with_backoff` and `roundtrip` return a `&[u8]`, which is the written-to portion of the rxbuf? This would avoid needing to index into the rxbuf correctly.

Reply:
I'll do a little research on this. But it looks like I left myself a little comment about this:

```
src/client/response.rs:

pub(crate) trait Response {
  // We leak the response length so that payloads can be parsed. This should be
  // cleaner when associated types within traits can specify named lifetime
  // parameters.
  fn deserialize(buf: &[u8]) -> Result<(Self, &[u8])>

```

Reply:
Woah—looks like this is easily doable now!!